File: rust_kernel/src/taskmaster/syscall/mknod.rs
use super::scheduler::SCHEDULER;
use super::vfs::{Path, VFS};
use super::SysResult;
use core::convert::TryFrom;
use libc_binding::{c_char, dev_t, mode_t, Errno, FileType};

pub fn sys_mknod(path: *const c_char, mut mode: mode_t, _dev: dev_t) -> SysResult<u32> {
    unpreemptible_context!({
        let mut scheduler = SCHEDULER.lock();

        let safe_path = {
            let v = scheduler
                .current_thread_mut()
                .unwrap_process_mut()
                .get_virtual_allocator();

            v.make_checked_str(path)?
        };
        let tg = scheduler.current_thread_group_mut();

        let mask = tg.umask;
        mode = mode & !mask;

        let creds = &tg.credentials;
        let cwd = &tg.cwd;
        let path = Path::try_from(safe_path)?;
        let mode = FileType::from_bits(mode as u16).ok_or(Errno::EINVAL)?;
        VFS.lock().mknod(cwd, creds, path, mode)?;
        Ok(0)
    })
}

File: libc/src/sys/stat/mknod.c
#include <sys/stat.h>
#include <user_syscall.h>
#include <errno.h>

#include <custom.h>

#warning NOT IMPLEMENTED

int mknod(const char *path, mode_t mode, dev_t dev)
{
	DUMMY_KERNEL
	int ret = _user_syscall(MKNOD, 3, path, mode, dev);
	set_errno_and_return(ret);
}
